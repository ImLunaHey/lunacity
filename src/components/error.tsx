import React from 'react';
import { Text } from '@nextui-org/react';

const ErrorComponent: React.FC<{ statusCode: number; message: string }> = ({
  message,
}) => {
  return (
    <div
      className="relative rounded border border-red-500 bg-red-500 px-4 py-3 text-white"
      role="alert"
    >
      <span className="block sm:inline">{message}</span>
    </div>
  );
  //   return <Text color="error">{message}</Text>;
};

export default ErrorComponent;
