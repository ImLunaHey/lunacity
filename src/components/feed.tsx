import React from 'react';
import { SinglePost } from './single-post';
import type { SinglePostProps } from './single-post';
import InfiniteScroll from 'react-infinite-scroll-component';
import { Loading } from '@nextui-org/react';
import { useTranslation } from 'react-i18next';

type FeedProps = {
  items?: SinglePostProps[];
  publicCursor?: string;
  personalCursor?: string;
  className?: string;
  fetchData?: (opts?: { personalCursor?: string; publicCursor?: string }) => () => void;
};

const getUniqueItemsInArray = (items: SinglePostProps[]): Exclude<SinglePostProps, null>[] => {
  const seen: Record<string, boolean> = {};
  return (items as Exclude<SinglePostProps, null>[]).filter((item) =>
    seen.hasOwnProperty(item.id) ? false : (seen[item.id] = true)
  );
};

const Feed: React.FC<FeedProps> = ({ items, fetchData = (opts) => () => undefined, personalCursor, publicCursor }) => {
  const { t } = useTranslation();
  if (!items) return null;
  const uniqueItems = getUniqueItemsInArray(items);

  return (
    <InfiniteScroll
      next={fetchData({ publicCursor, personalCursor })}
      dataLength={uniqueItems.length}
      hasMore={uniqueItems.length === 50}
      loader={<Loading />}
      endMessage={
        <p style={{ textAlign: 'center' }}>
          <b>{t('no-more-posts')}</b>
        </p>
      }
      // // below props only if you need pull down functionality
      // refreshFunction={fetchData()}
      // pullDownToRefresh
      // pullDownToRefreshThreshold={50}
      // pullDownToRefreshContent={
      //   <h3 style={{ textAlign: 'center' }}>&#8595; Pull down to refresh</h3>
      // }
      // releaseToRefreshContent={
      //   <h3 style={{ textAlign: 'center' }}>&#8593; Release to refresh</h3>
      // }
    >
      {uniqueItems.map((post) => (
        <SinglePost post={post} key={post.id} />
      ))}
    </InfiniteScroll>
  );
};

export default Feed;
