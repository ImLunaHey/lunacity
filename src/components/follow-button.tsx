'use client';

import { Button, Loading } from '@nextui-org/react';
import { type FC, useState } from 'react';
import { api } from '../utils/api';
import ErrorComponent from '../components/error';
import { useSession } from 'next-auth/react';
import { type TRPCClientErrorLike } from '@trpc/client';
import { useTranslation } from 'react-i18next';

type handleErrorProps = {
  error: TRPCClientErrorLike<any>;
  setError: (error: { statusCode: number; message: string } | null) => void;
};

const handleError = ({ error, setError }: handleErrorProps) => {
  setError({ statusCode: 500, message: error.message });
  setTimeout(() => {
    setError(null);
  }, 10_000);
};

export const FollowButton: FC<{ handle: string }> = ({ handle }) => {
  const { t } = useTranslation();
  const session = useSession();
  const followingState = api.page.followingState.useQuery({ handle }, { enabled: session.status === 'authenticated' });
  const followPage = api.page.followPage.useMutation();
  const unfollowPage = api.page.unfollowPage.useMutation();
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<{
    statusCode: number;
    message: string;
  } | null>(null);

  const onError = (error: TRPCClientErrorLike<any>) => {
    handleError({ error, setError });
    setIsLoading(false);
  };
  const onSuccess = () => {
    // @TODO: handle if this fails fetching
    void followingState.refetch().then(() => {
      setIsLoading(false);
    });
  };

  // @TODO: This needs a debounce
  const onPress = () => {
    setIsLoading(true);

    // Unfollow page
    if (followingState.data) {
      unfollowPage.mutate(
        { handle },
        {
          onError,
          onSuccess,
        }
      );
    } else {
      // Follow page
      followPage.mutate(
        { handle },
        {
          onError,
          onSuccess,
        }
      );
    }
  };

  const generateText = (isFollowing?: boolean) => {
    if (isFollowing === undefined) return;
    return isFollowing ? t('unfollow') : t('follow');
  };

  // If we hit an error show it
  if (error) return <ErrorComponent {...error} />;

  if (session.status === 'loading') return <Loading />;

  // Unauthenticated users cannot follow/unfollow
  if (session.status === 'unauthenticated') return null;

  // Show the follow button
  return (
    <Button
      className="min-w-fit"
      size="xs"
      color="secondary"
      disabled={isLoading}
      aria-label={generateText(followingState.data)}
      onPress={onPress}
    >
      {generateText(followingState.data)}
    </Button>
  );
};
