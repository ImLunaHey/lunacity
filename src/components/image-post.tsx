import type { SinglePostProps } from '@app/components/single-post';
import { TagCloud } from '@app/components/tag-cloud';
import { PageAvatar } from '@app/components/page-avatar';
import { Card, Grid, Button, Text } from '@nextui-org/react';
import { useTranslation } from 'react-i18next';

export const ImagePost: React.FC<{ post: SinglePostProps }> = ({ post }) => {
  const { t } = useTranslation();
  if (!post) return null;

  return (
    <Card css={{ p: '$6' }}>
      <Card.Header>
        <PageAvatar
          // @TODO: Fix this
          official={false}
          src={'https://nextui.org/images/card-example-5.jpeg'}
          name={post.page.displayName}
          handle={post.page.handle}
          description={post.page.description}
          followerCount={post.page.followerCount}
          followingCount={post.page.followingCount}
          popover={true}
        />
        <Grid.Container css={{ pl: '$6' }}>
          <Grid xs={12}>
            <Text h4>{post.page.displayName}</Text>
          </Grid>
        </Grid.Container>
      </Card.Header>
      <Card.Body css={{ py: '$2' }}>
        <div style={{ position: 'relative', width: '100%' }}>{/* <Image {...image} alt={image.alt} /> */}</div>
        {/* {body && (<ReactMarkdown remarkPlugins={[remarkGfm]}>{body}</ReactMarkdown>)} */}

        <Grid.Container alignItems="flex-start" gap={0.5}>
          <Grid>
            <TagCloud tags={post.tags} />
          </Grid>
        </Grid.Container>
      </Card.Body>
      <Card.Footer>
        <Grid.Container alignItems="flex-start" gap={0.5}>
          <Grid>
            <Button className="min-w-0" size="xs" color="default" aria-label={t('bookmark-post')}>
              üîñ
            </Button>
          </Grid>
          <Grid>
            <Button className="min-w-0" size="xs" color="default" aria-label={t('add-reaction')}>
              ‚≠ê
            </Button>
          </Grid>
          <Grid>
            <Button className="min-w-0" size="xs" aria-label={t('post-add-comment')}>
              üí¨
            </Button>
          </Grid>
        </Grid.Container>
      </Card.Footer>
    </Card>
  );
};
