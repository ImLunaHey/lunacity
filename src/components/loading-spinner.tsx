import type { FC } from 'react';
import { Loading } from '@nextui-org/react';

export const LoadingSpinner: FC = () => {
  return <Loading type="points-opacity" color="currentColor" size="sm" />;
};
