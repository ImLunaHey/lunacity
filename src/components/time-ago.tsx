import { humanTime } from '@app/common/human-time';
import { Badge } from '@nextui-org/react';
import type { FC } from 'react';
import { useTranslation } from 'react-i18next';

export const TimeAgo: FC<{ createdAt: Date; updatedAt?: Date }> = ({ createdAt, updatedAt }) => {
  const { t } = useTranslation();

  // If the updated time isn't the same as the created time then it's been updated
  const hasUpdated = createdAt.getTime() !== updatedAt?.getTime();
  const time = hasUpdated ? updatedAt : createdAt;

  // Don't render if we don't have at least one of the two "createdAt" or "updatedAt"
  if (!time) return null;

  return (
    <Badge isSquared>
      {hasUpdated ? t('Updated') : t('Posted')} {humanTime(time)}
    </Badge>
  );
};
