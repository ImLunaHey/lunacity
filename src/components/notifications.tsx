'use client';

import { LoadingSpinner } from '@app/components/loading-spinner';
import { useNotificationStore } from '@app/stores/notification';
import { Dropdown, Navbar, Badge, styled } from '@nextui-org/react';
import { useSession } from 'next-auth/react';
import { type FC, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { toast } from 'react-toastify';
import { NotificationIcon } from '../icons/notification-icon';
import { api } from '../utils/api';
import { FollowedNotification } from './notifications/followed-notification';

const NotificationButton = styled('button', {
  background: 'transparent',
  border: 'none',
  cursor: 'pointer',
  '&:active': {
    opacity: 0.8,
  },
});

export const Notifications: FC = () => {
  const { t } = useTranslation(['common']);
  const { data: session } = useSession();
  const [isLoading, setIsLoading] = useState(true);
  const notifications = useNotificationStore((state) => state.notifications);
  const addNotifications = useNotificationStore((state) => state.addNotifications);
  const addNotification = useNotificationStore((state) => state.addNotification);
  const handleSelection = () => undefined;

  // Get all the current notifications from the server
  api.notification.getAllNotifications.useQuery(
    {},
    {
      onSuccess(notifications) {
        addNotifications(notifications);
        setIsLoading(false);
      },
    }
  );

  api.notification.getLiveNotifications.useSubscription(undefined, {
    onData({ notification }) {
      // @TODO: This should be a jsx per type
      if (!notification) return;

      // Save the notification to the store
      addNotification(notification);

      if (notification.type === 'FOLLOWED')
        toast(<FollowedNotification notification={notification} />, { delay: 1000 });
    },
    onStarted() {
      if (process.env.NODE_ENV === 'development') toast.info(t('notifications.live-notifications-started'));
    },
    onError(error) {
      toast.error(error.message);
    },
    enabled: session?.user?.id !== undefined,
  });

  // Don't render unless the user has a page
  if (!session?.user?.page?.handle) return null;

  if (isLoading) return <LoadingSpinner />;

  if (notifications.length === 0)
    return (
      <Dropdown placement="bottom-right">
        <Navbar.Item>
          <Dropdown.Trigger>
            <NotificationButton aria-label={`${notifications.length} notifications`}>
              {/* @TODO: the content and above label should reflect the actual count, if above 99 show 99+ */}
              <Badge color="error" content={notifications.length} shape="circle" size="sm">
                <NotificationIcon fill="currentColor" width={30} height={25} />
              </Badge>
            </NotificationButton>
          </Dropdown.Trigger>
        </Navbar.Item>
        <Dropdown.Menu aria-label="User notifications" color="secondary">
          <Dropdown.Item
            key="no-notifications"
            className="h-fit items-start py-2 leading-normal text-gray-500 hover:text-white"
          >
            No notifications
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    );

  return (
    <Dropdown placement="bottom-right">
      <Navbar.Item>
        <Dropdown.Trigger>
          <NotificationButton aria-label={`${notifications.length} notifications`}>
            {/* @TODO: the content and above label should relfect the actual count, if above 99 show 99+ */}
            <Badge color="error" content={notifications.length} shape="circle" size="sm">
              <NotificationIcon fill="currentColor" width={30} height={25} />
            </Badge>
          </NotificationButton>
        </Dropdown.Trigger>
      </Navbar.Item>
      <Dropdown.Menu aria-label="User notifications" color="secondary" onSelectionChange={handleSelection}>
        {notifications.map((notification) => (
          <Dropdown.Item
            key={notification.id}
            // @TODO: Add different icons for each notification type
            icon={<NotificationIcon className="mt-2" size={16} />}
            className="h-fit items-start py-2 leading-normal text-gray-500 hover:text-white"
          >
            {notification.type === 'FOLLOWED' && <FollowedNotification notification={notification} />}
          </Dropdown.Item>
        ))}
      </Dropdown.Menu>
    </Dropdown>
  );
};
