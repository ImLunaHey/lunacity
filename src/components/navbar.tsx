import { useSSR } from '@nextui-org/react';
import React, { type FC } from 'react';
import { Navbar, Text, Input, Dropdown, Avatar } from '@nextui-org/react';
import { signOut, useSession } from 'next-auth/react';
import Link from 'next/link';
import { useRouter } from 'next/router';
import { SearchIcon } from '../icons/search-icon';
import { Notifications } from './notifications';
import { useTranslation } from 'react-i18next';
import { type SubmitHandler, useForm } from 'react-hook-form';
import Logo from '@app/components/logo';

const DropdownMenu: FC = () => {
  const { data: session } = useSession();
  const { t } = useTranslation(['common']);
  if (!session?.user?.page?.handle) {
    return (
      <Dropdown.Menu aria-label="User menu actions" color="secondary">
        <Dropdown.Item textValue="Create page" key="profile" className="p-0">
          <Link className="m-0 block h-full w-full p-2 text-white" href="/page/create">
            <Text b color="inherit" css={{ d: 'flex' }}>
              Create page
            </Text>
          </Link>
        </Dropdown.Item>
        <Dropdown.Item command="⌘S" textValue="Settings" key="settings" withDivider>
          <Link className="m-0 block h-full w-full text-gray-400 hover:text-white" href="/settings">
            {t('page.settings.title')}
          </Link>
        </Dropdown.Item>
        <Dropdown.Item textValue="Help and feedback" key="help_and_feedback">
          <Link className="m-0 block h-full w-full text-gray-400 hover:text-white" href="/help">
            {t('page.help.title')}
          </Link>
        </Dropdown.Item>
        <Dropdown.Item textValue="Signout" key="signout" withDivider color="error">
          <a
            className="m-0 block h-full w-full text-gray-400 hover:text-white"
            onClick={() => void signOut({ callbackUrl: '/' })}
          >
            {t('signout')}
          </a>
        </Dropdown.Item>
      </Dropdown.Menu>
    );
  }

  return (
    <Dropdown.Menu aria-label="User menu actions" color="secondary">
      <Dropdown.Item textValue={`@${session.user?.page?.handle}`} key="profile" className="p-0">
        <Link className="m-0 block h-full w-full p-2 text-white" href={`/${session.user?.page?.handle}`}>
          <Text b color="inherit" css={{ d: 'flex' }}>
            @{session.user?.page?.handle}
          </Text>
        </Link>
      </Dropdown.Item>
      <Dropdown.Item command="⌘S" textValue="Settings" key="settings" withDivider>
        <Link className="m-0 block h-full w-full text-gray-400 hover:text-white" href="/settings">
          {t('page.settings.title')}
        </Link>
      </Dropdown.Item>
      <Dropdown.Item textValue="Analytics" key="analytics">
        <Link className="m-0 block h-full w-full text-gray-400 hover:text-white" href="/analytics">
          {t('page.analytics.title')}
        </Link>
      </Dropdown.Item>
      <Dropdown.Item textValue="Help and feedback" key="help_and_feedback">
        <Link className="m-0 block h-full w-full text-gray-400 hover:text-white" href="/help">
          {t('page.help.title')}
        </Link>
      </Dropdown.Item>
      <Dropdown.Item textValue="Signout" key="signout" withDivider color="error">
        <a
          className="m-0 block h-full w-full text-gray-400 hover:text-white"
          onClick={() => void signOut({ callbackUrl: '/' })}
        >
          {t('signout')}
        </a>
      </Dropdown.Item>
    </Dropdown.Menu>
  );
};

const UserMenu: FC = () => {
  return (
    <Dropdown placement="bottom-right">
      <Navbar.Item>
        <Dropdown.Trigger>
          <Avatar
            bordered
            as="button"
            color="primary"
            size="md"
            src="https://i.pravatar.cc/150?u=a042581f4e29026704d"
          />
        </Dropdown.Trigger>
      </Navbar.Item>
      <DropdownMenu />
    </Dropdown>
  );
};

export default function NavBar() {
  const { data: session } = useSession();
  const router = useRouter();
  const { t } = useTranslation(['common']);
  const { isBrowser } = useSSR();

  // Search box
  type Inputs = { query: string };
  const { register, handleSubmit } = useForm<Inputs>();
  const onSubmit: SubmitHandler<Inputs> = (data) => {
    void router.push({
      pathname: '/search',
      query: {
        query: data.query,
      },
    });
  };

  // Don't render the navbar when we're within SSR
  // See: https://github.com/nextui-org/nextui/issues/779
  if (!isBrowser) return null;

  // If there's no session at all don't render
  if (!session) return null;

  return (
    <Navbar isBordered variant="sticky">
      <Navbar.Brand css={{ mr: '$4' }}>
        <Logo />
        <Link href="/">
          <Text b color="inherit" css={{ mr: '$11' }} hideIn="xs">
            Lunacity
          </Text>
        </Link>
        <Navbar.Content hideIn="xs" variant="highlight">
          <Navbar.Link isActive={router.asPath === '/'} href="/" as={Link}>
            🏡 {t('page.home.title')}
          </Navbar.Link>
          <Navbar.Link isActive={router.asPath === '/explore'} href="/explore" as={Link}>
            🌏 {t('page.explore.title')}
          </Navbar.Link>
          <Navbar.Link isActive={router.asPath === '/messages'} href="/messages" as={Link}>
            📩 {t('page.messages.title')}
          </Navbar.Link>
        </Navbar.Content>
      </Navbar.Brand>

      {/* Search box */}
      <Navbar.Content
        className="w-full"
        css={{
          ':first-child': {
            width: '100%',
          },
        }}
      >
        <Navbar.Item className="mr-3 w-full">
          {/* eslint-disable-next-line @typescript-eslint/no-misused-promises */}
          <form onSubmit={handleSubmit(onSubmit)}>
            <Input
              aria-label="Search box"
              clearable
              contentLeft={<SearchIcon fill="var(--nextui-colors-accents6)" size={16} />}
              contentLeftStyling={false}
              className="w-full"
              css={{
                '& .nextui-input-content--left': {
                  height: '100%',
                  marginLeft: '$4',
                  dflex: 'center',
                },
              }}
              placeholder="Search..."
              {...register('query', { required: true, maxLength: 25 })}
            />
          </form>
        </Navbar.Item>
      </Navbar.Content>
      <Navbar.Content
        css={{
          '@xsMax': {
            jc: 'space-between',
          },
        }}
      >
        <Notifications />
        <UserMenu />
      </Navbar.Content>
    </Navbar>
  );
}
