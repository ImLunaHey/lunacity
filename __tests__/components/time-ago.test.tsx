import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { TimeAgo } from '../../src/components/time-ago';

jest.mock('react-i18next', () => ({
  useTranslation: jest.fn(() => ({
    t: jest.fn((text: string) => text),
    ready: true,
  })),
}));

describe('TimeAgo', () => {
  it('renders createdAt time if no updatedAt is provided', () => {
    const { container } = render(<TimeAgo createdAt={new Date()} />);
    expect(container).toMatchSnapshot();
  });

  it('renders createdAt time if no updatedAt is the same as createdAt', () => {
    const date = new Date();
    const { container } = render(<TimeAgo createdAt={date} updatedAt={date} />);
    expect(container).toMatchSnapshot();
  });

  it('renders updatedAt time if no updatedAt is different to createdAt', () => {
    const createdAt = new Date();
    const updatedAt = new Date(new Date().getTime() + 10_000);
    const { container } = render(<TimeAgo createdAt={createdAt} updatedAt={updatedAt} />);
    expect(container).toMatchSnapshot();
  });
});
