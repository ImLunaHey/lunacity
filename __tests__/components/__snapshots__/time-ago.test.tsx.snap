// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAgo renders createdAt time if no updatedAt is provided 1`] = `<div />`;

exports[`TimeAgo renders createdAt time if no updatedAt is the same as createdAt 1`] = `
<div>
  <span
    class="nextui-c-gyluqH nextui-badge-root"
  >
    <span
      class="nextui-c-jTSrtu nextui-c-jTSrtu-jCnmOD-size-md nextui-c-jTSrtu-dKqOOw-color-default nextui-c-jTSrtu-UlDvZ-isSquared-true nextui-c-jTSrtu-iCGBCn-placement-top-right nextui-c-jTSrtu-foBxhG-borderWeight-normal nextui-c-jTSrtu-bnEBqI-disableOutline-false nextui-c-jTSrtu-iXycSg-disableAnimation-true nextui-badge"
    >
      posted
       
      now
    </span>
  </span>
</div>
`;

exports[`TimeAgo renders updatedAt time if no updatedAt is different to createdAt 1`] = `
<div>
  <span
    class="nextui-c-gyluqH nextui-badge-root"
  >
    <span
      class="nextui-c-jTSrtu nextui-c-jTSrtu-jCnmOD-size-md nextui-c-jTSrtu-dKqOOw-color-default nextui-c-jTSrtu-UlDvZ-isSquared-true nextui-c-jTSrtu-iCGBCn-placement-top-right nextui-c-jTSrtu-foBxhG-borderWeight-normal nextui-c-jTSrtu-bnEBqI-disableOutline-false nextui-c-jTSrtu-iXycSg-disableAnimation-true nextui-badge"
    >
      updated
       
      now
    </span>
  </span>
</div>
`;
